File: app/src/main/java/jm/through/read/MessageActivity.java
package jm.through.read;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.sun.mail.iap.Argument;

import org.w3c.dom.Text;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import javax.activation.DataHandler;
import javax.mail.BodyPart;
import javax.mail.Multipart;
import javax.mail.Part;
import javax.mail.internet.MimeUtility;

import jm.through.R;

import static jm.through.read.FolderFetchImap.readList;
//import static jm.through.read.MailReader.readList;

/**
 * 메일 상세정보 확인창 작업중
 */
public class MessageActivity extends AppCompatActivity {
    // 초기화해야하는 리스트를 초기화하는 곳


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case (android.R.id.home): {
                finish();
                break;
            }
        }
        return true;
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_content);
        Toolbar t = findViewById(R.id.message_bar);
        setSupportActionBar(t);
        getSupportActionBar().setTitle("제목 히히");
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        final TextView titleString = (TextView) findViewById(R.id.message_content_title);
        final TextView memoString = (TextView) findViewById(R.id.message_content_memo);
        final TextView dateString = (TextView) findViewById(R.id.message_content_time);

        //final TextView itemString = (TextView)v.findViewById(R.id.)
        SimpleDateFormat df = new SimpleDateFormat("yyyy.MM.dd EE요일, aa hh:mm");

        if (getIntent() != null) {
//            Bundle bundle = getArguments();
//            int a = bundle.getInt("position");
//
//            System.out.println("bundle is =" +bundle);

            int a = getIntent().getIntExtra("position", 0);

            String getSubject = readList.get(a).component1().split("<")[0]; // subject
            String getFrom = readList.get(a).component2(); // from
            Date getDate = readList.get(a).component3(); // date
            // ImageView getimage= readList.get(a).component4();
            //System.out.println("subject printing... "+ getSubject.toString());

            String mTime = df.format(getDate);

//            // 추가 함
//
//            DataHandler dataHandler = bundle.getDataHandler();
//            String contentType = dataHandler.getContentType();
//
//            if(contentType.indexOf("multipart/mixed") != -1){
//                Multipart multipart = (Multipart)msg.getContent();// javax.mail.internet.MimeMultipart@8d463af
//                try{
//                    //System.out.println("multipart.getCount():"+multipart.getCount());
//                    for(int j=0;j<multipart.getCount();j++){
//                        BodyPart part = multipart.getBodyPart(j);
//                        //System.out.println(part.getContent());
//                        String disp = part.getDisposition();
//                        if (disp == null || disp.equalsIgnoreCase(Part.ATTACHMENT)){
//                            //첨부 파일 이라는 얘기임
//
//                            String filename =part.getFileName();
//
//                            System.out.println("첨부된 파일 이름 ="+filename);
//
////                                if (filename.startsWith("=?ks_c_5601-1987?B?")) //BASE64 로 인코딩 되있는 경우
////                                {
////                                    filename = filename.substring(19) ;
////                                    if (filename.indexOf('?') != -1) filename = filename.substring( 0 ,filename.indexOf('?') ) ;
////                                    filename = new String(MimeUtility.decodeText(filename)) ;
////                                }else if(filename.startsWith("=?ISO-8859-1?B?")){
////                                    filename = filename.substring(15) ;
////                                    if (filename.indexOf('?') != -1) filename = filename.substring( 0 ,filename.indexOf('?') ) ;
//                            filename = new String(MimeUtility.decodeText(filename)) ;
//                            Log.v("파일명 ","파일명:"+filename);
////                                }
//
//
//                            try{
//
//                                System.out.println("첨부된 파일 이름[인코딩]:"+ filename);
//                                File f = new File("C:"+File.separator+"Temp"+File.separator+"mail"+File.separator+filename);
//                                FileOutputStream fos = new FileOutputStream(f);
//
//                                byte[] buffer = new byte[1024];
//
//                                BufferedInputStream in = new BufferedInputStream(part.getInputStream());
//                                Log.v("여기야!! InputStream = ",in.toString());
//
//                                int n = 0;
//                                while((n=in.read(buffer, 0, 1024)) != -1) {
//                                    fos.write(buffer, 0, n);
//                                }
//                                fos.close();
//                                in.close();
//                            }catch(Exception e1){
//                                System.out.println("[SUB_ERROR]:"+e1.getMessage());
//                            }
//                        }//end attech
//
//
//                        //System.out.println("["+disp+"]"+part.getContentType());
//                    }
//                }catch(Exception e){
//                    System.out.println("[ERROR]:"+e.getMessage());
//                }
//            }
//
//
//            if ( (contentType.indexOf("text/html") != -1) || (contentType.indexOf("text/plain") != -1) ) {
//                BufferedReader br = new BufferedReader(new InputStreamReader(dataHandler.getInputStream()));
////                    char[] buff = new char[512];
////                    int len;
////                    while ( (len = br.read(buff)) != -1) {
////                        System.out.print(new String(buff, 0, len));
////                    }
//                Log.v("여기야!! reading html...",br.toString());
//                br.close();
//            }
//            System.out.println("\n=====================================================\n");


            dateString.setText(mTime);
            titleString.setText(getSubject);
            memoString.setText(getFrom);
            //   imageView.setImageURI();
        }
    }

    // 레이아웃을 inflate 하고 각 리소스를 초기화하는 곳

//    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
//        View v = inflater.inflate(R.layout.fragment_content,container,false);
//
//
//
//        return v;
//    }

    public void onBackPressed() {
        finish(); //액티비티 종료
//       if(getFragmentManager().getBackStackEntryCount() > 0){
//           getFragmentManager().popBackStack();
//       }
    }
}

File: app/src/main/java/jm/through/read/MessageActivity.kt
package jm.through.read

import android.content.Context
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import android.widget.TextView
import android.widget.Toast
import jm.through.R
import jm.through.read.FolderFetchImap.readList
import java.io.IOException
import java.text.SimpleDateFormat
import javax.mail.BodyPart
import javax.mail.MessagingException
import javax.mail.Multipart
import javax.mail.internet.ContentType
import javax.mail.internet.MimeMessage
import javax.mail.internet.MimeMultipart
import javax.mail.internet.MimeUtility


/*
메일 상세보기 창
 */


class MessageActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_content);
        Log.v("Loading.. ", "MessageActivity!");

        /*
        boundary = 본문의 서로 다른 부분을 구분하기 위한 구분자로 쓰임
        본문에서 boundary 뽑고 따로 읽기

        */
//        // 여기서 에러남
//        val t = findViewById<Toolbar>(R.id.message_bar)
//        setSupportActionBar(t)
//        supportActionBar!!.setTitle("MessageActivity 툴바!!")
//        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        Log.v("messageActivity - ", "toolbar ");
        val titleString = findViewById(R.id.message_content_title) as TextView // 제목
        val memoString = findViewById(R.id.message_content_memo) as TextView // 내용(Only text)
        val dateString = findViewById(R.id.message_content_time) as TextView // 시각
        val mWebView = findViewById(R.id.message_content_web) as WebView // 내용 content

        val webSettings = mWebView.getSettings()
        webSettings.setJavaScriptEnabled(true)
        webSettings.defaultTextEncodingName = "UTF-8"

        val df = SimpleDateFormat("yyyy.MM.dd EE요일, aa hh:mm")

        Log.v("messageActivity -", "getIntent()");
        if (getIntent() != null) {
            val intent = getIntent();
            val a = intent.getIntExtra("position", 0);
            Log.v("intent ", " is " + a);
            val getSubject = readList[a].mailTitle.split("<".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0] // subject
            val getFrom = readList[a].mailMemo // from
            val getDate = readList[a].mailDate // date
            val getContents = readList[a].mailContent
            val getContenttype = readList[a].mailContenttype


            //val mp = readList[a].mailContent as Multipart // test
            val mTime = df.format(getDate)

            dateString.text = mTime
            titleString.text = getSubject
            memoString.text = getFrom

            System.out.println("mail's contenttype : " + getContenttype);


            @Throws(MessagingException::class, IOException::class)
            fun getTextFromMimeMultipart(mimeMultipart:MimeMultipart):String{
                val count = mimeMultipart.count
                if (count == 0)
                    throw MessagingException("Multipart with no body parts not supported.")
                val multipartAlt = ContentType(mimeMultipart.contentType).match("multipart/alternative")

                fun getTextFromBodyPart(bodyPart: BodyPart):String{
                    var result = ""

                    try {
                        if (bodyPart.isMimeType("text/plain") || bodyPart.isMimeType("text/html")) {
                            result = bodyPart.content as String
                        }
//                        else if (bodyPart.isMimeType("text/html")) {
//                            result = bodyPart.content as String}
                        else if (bodyPart.content is MimeMultipart){
                            result = getTextFromMimeMultipart(bodyPart.content as MimeMultipart)
                        }
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                    return MimeUtility.decodeText(result)
                }

                // Multipart ERROR
                if (multipartAlt)
                    return getTextFromBodyPart(mimeMultipart.getBodyPart(count - 1))
                var result = ""
                for (i in 0 until count) {
                    val bodyPart = mimeMultipart.getBodyPart(i)
                    // 실행 안됨
                    Log.v("running"+i.toString()," 번째 도는 중..")
                    result += getTextFromBodyPart(bodyPart)
                }
                return MimeUtility.decodeText(result)
            }

            @Throws(MessagingException::class, IOException::class)
            fun getTestFromMessage(message:Object) : String{
                var result = ""
                if (getContenttype.contains("text")) {
                    result = message.toString()
                    Log.v("multipart debug","text decoding")

                    //TODO Comment by jiran
                    /** contains는 문구가 포함됬는지를 확인하는 api임.
                    contents-type이 multipart/alternative일 때 아래 분기문이 스킵되면서 컨텐츠를 획득하지 않음.

                    + 질문 있습니다!! multipart로 구문을 찾는데 multipart/alternative 인 경우 분기문이 어떤 이유로
                    스킵되는 건가요..?
                     */
                } else if (getContenttype.contains("multipart", true)) {
                    val mimeMultipart = message as MimeMultipart
                    result = getTextFromMimeMultipart(mimeMultipart)
                    Log.v("multipart debug","multipart decoding");
                }

                return MimeUtility.decodeText(result)
            }

            /**
             * HTML URL 할 것
             * 순서 1. 주소 클릭 시 바로 이동하지 않고 이동할 주소 띄우기
             *      방법 1. a href 앞에 새로운 a href를 넣어 이동할 주소를 묻는 걸 띄워보기
             * 순서 2. Yes 시 이동 O, No 시 이동 X
             * 순서 3. 주소 클릭 시 바로 이동하지 않고 이동할 주소 오픈 그래프로 띄우기
             *
             * 아래 = 웹페이지 용 마우스를 링크에 가져다대면 상태표시줄에 링크 표
             * <a href="http://aaa.com" onclick="window.open(this.href, '', ''); return false;">
             *
             * 테스트 해볼 것
             * 아래 = href에서 발생한 액션을 실행시키지 않는 = 슈도프로토
             * <a href="#" onclick="return false;">test</a>
             *
             * <a onclick="myFunc();">Hello</a>
             *
             * $(document).ready(function(){ $('a').css('cursor','pointer');});
             *
             * href 를 찾아 onClick속성을 추가시켜 커스텀 함수로(링크를 보여주는 곳) 이동시켜야함
             * onclick="callFunction(); return false;
             */

            fun JavaScriptParser(mContext: Context) {
                @JavascriptInterface
                fun showToast(toast: String) {
                    Toast.makeText(mContext, toast, Toast.LENGTH_SHORT).show()
                }
            }

          //  mWebView.addJavascriptInterface(JavaScriptPasser(this), "Android")

            //TODO Comment by jiran
            /* 메일 파싱 부분은 Network이기 때문에 별도의 WorkingThread로 변경
            (별도 스레드로 추출하지 않으면, NetworkOnMainThreadException 발생.) */
            Thread(Runnable {
                var temp = getTestFromMessage(getContents)

//                val count = StringUtils.countOccurrences(temp, "<a href")
                //TODO Comment by jiran
                /* UI 컴포넌트의 API는 메인스레드(UI)에서만 요청할 수 있음
                (GUI 프로그래밍은 동기화 문제 방지를 위해 모두 같은 개념을 가짐.) */
                runOnUiThread({ mWebView.loadData(temp, "text/html", "UTF-8"); })
            }).start()
        }
    }


    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }

}

